src/Features/Core/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -19,6 +19,14 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     internal abstract partial class AbstractSymbolCompletionProvider : AbstractCompletionProvider
     {
+        // PERF: Many CompletionProviders dervice AbstractSymbolCompletionProvider and therefore
+        // compute identical contexts. This actually shows up on the 2-core typing test.
+        // Cache the most recent document/position/computed SyntaxContext to reduce repeat computation.
+        private static Document cachedDocument;
+        private static int cachedPosition;
+        private static AbstractSyntaxContext cachedContext;
+        private static readonly object cacheGate = new object();
```

Comment:
@pharring, I believe you're referring src\Features\VisualBasic\Completion\KeywordRecommenders\Types\BuiltInTypesKeywordRecommender.vb, right? I'll follow that pattern.

@dpoeschl That sounds like a good approach.